File: web/src/Repo/GroupRepo.js
import {fetchWrapper} from './browserWrapper'

export default class GroupRepo {
  baseUrl = 'http://localhost:4000/api'
  headers = {
    'Accept': 'application/json',
    'Origin': 'http://localhost:3000'
  }

  getList() {
    return fetchWrapper(`${this.baseUrl}/groups`, {headers: this.headers})
      .then(response => response.json())
      .then(json => json.data)
  }

  getDetails(groupId) {
    return fetchWrapper(`${this.baseUrl}/groups/${groupId}`, {headers: this.headers})
      .then(response => response.json())
      .then(json => json.data)
  }

  create(newGroup) {
    const url = `${this.baseUrl}/groups`

    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        'Origin': 'http://localhost:3000'
      },
      body: JSON.stringify({group: newGroup})
    }

    return fetchWrapper(url, options)
      .then(response => response.json())
      .then(json => json.data)
  }
}

File: ios/Chatter/Repo/GroupRepo.swift
import Foundation
import Combine

final class GroupRepo {
    func getList() -> AnyPublisher<[Group], ChatterError> {
        guard let url = URL(string: "http://localhost:4000/api/groups") else {
            return Fail(error: ChatterError()).eraseToAnyPublisher()
        }
        
        var request = URLRequest(url: url)
        request.httpBody = nil
        request.httpMethod = "GET"
        request.addValue("application/json", forHTTPHeaderField: "Accept")

        let session = URLSession.shared
        
        return session.dataTaskPublisher(for: request)
            .mapError { error in
                ChatterError()
            }
            .flatMap(maxPublishers: .max(1)) { pair in
                self.decode(pair.data)
            }
        .eraseToAnyPublisher()
    }
    
    private func decode(_ data: Data) -> AnyPublisher<[Group], ChatterError> {
        let decoder = JSONDecoder()
                
      return Just(data)
        .decode(type: ApiResponse.self, decoder: decoder)
        .map { response in
            response.data
        }
        .mapError { error in
            ChatterError()
        }
        .eraseToAnyPublisher()
    }
}
